### ISSUE TITLE
use safeParse to convert address to BigInt

### ISSUE THREAD

**Issue by Yelvabaoji** [shamblett/ethereum#5]
As Ethereum address may start from 0x0..., therefore, use safeParse to convert address to BigInt will lost the leading 0. After that, when convert the BigInt back to address, the beginning 0 will be eat.
This happen during call the "getBalance" function. For example, in /test/ethereum_common.dart, when call client.eth.getBalance, the address is convert from String to BigInt, and inside the getBalance, this same BigInt is convert to String, unfortunately, the leading 0 is cut during conversion.
By the way, "Future<int> getBalance" return an int, which will be wrong when the balance is big enough., so here return BigInt, i.e. "Future<BigInt> getBalance"

**Comment by shamblett** [comment]
Yes, there's no good way I can see to preserve this without introducing an EthereumQuantity class that contains both the raw string representation and the BigInt, allowing users to pick the one they want, unless you can suggest an alternative.

**Comment by gkd720** [comment]
After getting past my other issue(#4), I now see this problem. I have an account/address with a leading zero. Sending a transaction to geth results in an error:
```
ERROR::personal_sendTransaction - Code : -32602 <> Message : invalid argument 0: json: cannot unmarshal hex string of odd length into Go struct field SendTxArgs.from of type common.Address <> Id : 8
```

I'm calling it with a standard style (I think):

```
String currentAddress = "";
. . .
BigInt hashTxCreate = await client.admin.personalSendTransaction(BigInt.parse(currentAddress), currentPassword, . . .
```

Any ideas on a simple workaround?

I guess one idea could be to always round up these kinds of strings to an even length, but this wouldn't be a complete fix, since 2 or more leading zeros would still be bad, although less likely.

Another might be to keep the fields (from, to, data) as Strings. Do they have to be BigInts? I suspect signing/hashing might do some actual numerical manipulation.

Or, when turning a BigInt into a from/to string, always pad out with zeros to the 160 bit/40 hex length. Of course, this doesn't work for 'data'.

Or, add a 'pad' argument to 'bigIntegerToHex'.

Oh well, just doing some brainstorming. Thanks for looking into this.

**Comment by shamblett** [comment]
Yes, I need to look at this, BigInt seems the way to go rather than strings but it does have to work of course.  I was thinking of extending BigInt and creating an EthereumBigInt that handles all this for you but its not as simple as I envisaged, you can't just extend BigInt I'd have to wrap it  somehow. Other projects must have come across this before, Ill have a look round and come up with something

**Comment by gkd720** [comment]
Great. Thanks. I continued my development using an address without a leading zero, and I was able to complete a transaction, and find it recorded into a block. But I now have a concern over the "data" field  in a transaction. You have "data" as a BigInt, and I'm currently assigning it a number, so this works. But I envision using this field to save a text string, so I don't think this could always be easily parsed into  a BigInt due to spaces, punctuation, etc. From the Ethereum Whitepaper/spec, it describes "data" as _An unlimited size byte array specifying the input data of the message call, formally Td._  So is your implementation as a BigInt too restrictive? Or would I just have to turn any text string I plan on saving into a byte string myself first (any problems with Unicode?), and then turn that into a BigInt?

**Comment by shamblett** [comment]
OK, looking at this and taking into account the comments above the only sane way to proceed here is to create new type classes for the address and data types and use these instead of BigInt. This is a large change to the API but needs to be done and means we can get rid of the safe parse and others form Utilities. I'm working on this now and hope to release a new client say mid/end next week.

**Comment by shamblett** [comment]
Finally, we have two new datatypes EthereumAddress and EthereumData, this cleans up the API greatly, client re-published at 4.0.0
